File: src/main/java/com/ecommerce/modelo/Pedido.java
package com.ecommerce.modelo;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Pedido {
    private final int id;
    private final Cliente cliente;
    private final LocalDateTime fechaCreacion;
    private EstadoPedido estado;
    private final List<DetallePedido> detalles;

    public Pedido(int id, Cliente cliente, List<DetallePedido> detalles) {
        this.id = id;
        this.cliente = cliente;
        this.fechaCreacion = LocalDateTime.now();
        this.estado = EstadoPedido.PENDIENTE;
        this.detalles = new ArrayList<>(detalles);
    }

    public int getId() {
        return id;
    }

    public Cliente getCliente() {
        return cliente;
    }

    public EstadoPedido getEstado() {
        return estado;
    }

    public void cambiarEstado(EstadoPedido estado) {
        this.estado = estado;
    }

    public double calcularTotal() {
        return detalles.stream()
                .mapToDouble(DetallePedido::calcularSubtotal)
                .sum();
    }

    public LocalDateTime getFechaCreacion() {
        return fechaCreacion;
    }

    public List<DetallePedido> getDetalles() {
        return Collections.unmodifiableList(detalles);
    }
}

File: internal/ecommerce/pedido.go
package ecommerce

import (
	"fmt"
	"time"
)

type EstadoPedido string

const (
	EstadoPendiente EstadoPedido = "pendiente"
	EstadoPagado    EstadoPedido = "pagado"
	EstadoEnviado   EstadoPedido = "enviado"
	EstadoCancelado EstadoPedido = "cancelado"
)

type DetallePedido struct {
	productoID string
	nombre     string
	cantidad   int
	precio     float64
}

func (d DetallePedido) ProductoID() string { return d.productoID }
func (d DetallePedido) Nombre() string     { return d.nombre }
func (d DetallePedido) Cantidad() int      { return d.cantidad }
func (d DetallePedido) Precio() float64    { return d.precio }
func (d DetallePedido) Subtotal() float64  { return d.precio * float64(d.cantidad) }

type Pedido struct {
	id        string
	clienteID string
	estado    EstadoPedido
	total     float64
	fecha     time.Time
	detalles  []DetallePedido
}

func NewPedido(id string, cliente Cliente, carrito Carrito, inventario *Inventario) (*Pedido, error) {
	if carrito.Vacio() {
		return nil, ErrCarritoVacio
	}
	detalles := make([]DetallePedido, 0, len(carrito.Items()))
	for _, item := range carrito.Items() {
		producto, err := inventario.BuscarProducto(item.Producto().ID())
		if err != nil {
			return nil, err
		}
		if err := producto.ReducirStock(item.Cantidad()); err != nil {
			return nil, err
		}
		detalles = append(detalles, DetallePedido{
			productoID: producto.ID(),
			nombre:     producto.Nombre(),
			cantidad:   item.Cantidad(),
			precio:     producto.Precio(),
		})
	}
	return &Pedido{
		id:        id,
		clienteID: cliente.ID(),
		estado:    EstadoPendiente,
		total:     carrito.CalcularTotal(),
		fecha:     time.Now(),
		detalles:  detalles,
	}, nil
}

func (p Pedido) ID() string                { return p.id }
func (p Pedido) ClienteID() string         { return p.clienteID }
func (p Pedido) Estado() EstadoPedido      { return p.estado }
func (p Pedido) Total() float64            { return p.total }
func (p Pedido) Fecha() time.Time          { return p.fecha }
func (p Pedido) Detalles() []DetallePedido { return append([]DetallePedido{}, p.detalles...) }

func (p *Pedido) CambiarEstado(estado EstadoPedido) error {
	switch estado {
	case EstadoPendiente, EstadoPagado, EstadoEnviado, EstadoCancelado:
		p.estado = estado
		return nil
	default:
		return fmt.Errorf("estado %s: %w", estado, ErrEstadoInvalido)
	}
}

func (p *Pedido) Cancelar() error {
	if p.estado == EstadoEnviado {
		return fmt.Errorf("pedido %s enviado: no se puede cancelar", p.id)
	}
	p.estado = EstadoCancelado
	return nil
}
